loadbalancer/consul/publisher.go

```diff
@@ -0,0 +1,175 @@
+package consul
+
+import (
+	"fmt"
+
+	"github.com/go-kit/kit/endpoint"
+	"github.com/go-kit/kit/loadbalancer"
+	"github.com/go-kit/kit/log"
+	consul "github.com/hashicorp/consul/api"
+)
+
+const defaultIndex = 0
+
+// Publisher yields endpoints for the given service filtered by tags.
+type Publisher struct {
+	cache      *loadbalancer.EndpointCache
+	client     Client
+	logger     log.Logger
+	service    string
+	tags       []string
+	endpointsc chan []endpoint.Endpoint
+	quitc      chan struct{}
+}
+
+// NewPublisher returns a Consul publisher which returns Endpoints for the
+// requested service. It only returns instances for which all of the passed
+// tags are present.
+func NewPublisher(
+	client Client,
+	factory loadbalancer.Factory,
+	logger log.Logger,
+	service string,
+	tags ...string,
+) (*Publisher, error) {
+	logger = log.NewContext(logger).With("component", "Consul Publisher")
+
+	p := &Publisher{
+		cache:   loadbalancer.NewEndpointCache(factory, logger),
+		client:  client,
+		logger:  logger,
+		service: service,
+		tags:    tags,
+		quitc:   make(chan struct{}),
+	}
+
+	is, index, err := p.getInstances(defaultIndex)
+	if err != nil {
+		return nil, err
+	}
+
+	p.cache.Replace(is)
+
+	go p.loop(index)
+
+	return p, nil
+}
+
+// Endpoints implements the Publisher interface.
+func (p *Publisher) Endpoints() ([]endpoint.Endpoint, error) {
+	return p.cache.Endpoints(), nil
+}
+
+// Stop terminates the publisher.
+func (p *Publisher) Stop() {
+	close(p.quitc)
+}
+
+func (p *Publisher) loop(
+	lastIndex uint64,
+) {
+	var (
+		errc = make(chan error)
+		resc = make(chan response)
+	)
+
+	for {
+		go func() {
+			is, index, err := p.getInstances(lastIndex)
+			if err != nil {
+				errc <- err
+				return
+			}
+
+			resc <- response{
+				index:     index,
+				instances: is,
+			}
+		}()
+
+		select {
+		case err := <-errc:
+			p.logger.Log("service", p.service, "err", err)
+		case res := <-resc:
+			p.cache.Replace(res.instances)
+
+			lastIndex = res.index
+		case <-p.quitc:
+			return
+		default:
+
+		}
+	}
+}
+
+func (p *Publisher) getInstances(lastIndex uint64) ([]string, uint64, error) {
+	tag := ""
+
+	if len(p.tags) > 0 {
+		tag = p.tags[0]
+	}
+
+	entries, meta, err := p.client.Service(
+		p.service,
+		tag,
+		&consul.QueryOptions{
+			WaitIndex: lastIndex,
+		},
+	)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	if len(p.tags) > 1 {
+		entries = filterEntries(entries, p.tags[1:]...)
+	}
+
+	return makeInstances(entries), meta.LastIndex, nil
+}
+
+// response is used as container to transport instances as well as the updated
+// index.
+type response struct {
+	index     uint64
+	instances []string
+}
+
+func filterEntries(entries []*consul.ServiceEntry, tags ...string) []*consul.ServiceEntry {
+	es := []*consul.ServiceEntry{}
+
+ENTRIES:
+	for _, entry := range entries {
+		fmt.Println(entry.Service.ID)
+		tagMap := map[string]struct{}{}
```

Comment:
This is actually a set, and we know its exact size: `tagSet := make(map[string]struct{}, len(entry.Service.Tags))`